common/optimiser/Optimiser.java

```diff
@@ -87,23 +83,32 @@ public boolean isError() {
         return status == Status.ERROR;
     }
 
-    private boolean constraintsChanged() {
-        return iterate(constraints).anyMatch(c -> !c.isUpToDate);
+    public synchronized void reset() {
+        if (hasSolver) {
+            solver.reset();
+            setSolverModel();
+        } else initialiseSolver();
     }
 
     private void initialiseSolver() {
-        solver = MPSolver.createSolver("SAT");
+        solver = MPSolver.createSolver("SCIP");
```

Comment:
SAT is a faster solver overall, but will occasionally take 800 ms over the time limit before stopping. SCIP respects the time limit to within a few ms, but is orders of magnitude slower to converge.

Reply:
we should keep SAT for now then